gnosis/eth/ethereum_client.py

```diff
@@ -1066,61 +1066,37 @@ def batch_call(self, contract_functions: Iterable[ContractFunction], from_addres
 
         return self.batch_call_custom(payloads, raise_exception=raise_exception, block_identifier=block_identifier)
 
-    def estimate_gas(self, from_: str, to: str, value: int, data: EthereumData,
+    def estimate_gas(self, to: str, from_: Optional[str] = None, value: Optional[int] = None,
+                     data: Optional[EthereumData] = None, gas: Optional[int] = None, gas_price: Optional[int] = None,
                      block_identifier: Optional[BlockIdentifier] = 'latest'):
         """
         Workaround to support 'pending' `block_identifier` for Geth
         :param from_:
         :param to:
         :param value:
         :param data:
-        :param block_identifier:
+        :param gas:
+        :param gas_price:
+        :param block_identifier: Be careful, Geth does not support `pending` when estimating
         :return:
         """
-        data = HexBytes(data) if data else HexBytes(b'')
-        params: List[Union[Dict[str, Any], BlockIdentifier]] = [
-            {"from": from_,
-             "to": to,
-             "data": data.hex(),
-             "value": "0x{:x}".format(value),  # No leading zeroes
-             },
-        ]
-
-        if block_identifier:
-            params.append(block_identifier)
-
-        payload = {
-            "method": "eth_estimateGas",
-            "params": params,
-            "jsonrpc": "2.0",
-            "id": 1
-        }
-
-        response = self.http_session.post(url=self.ethereum_node_url, json=payload)
-        response_json = response.json()
-        if 'error' in response_json:
-            # When using `pending`, Geth returns
-            """
-            {
-                "jsonrpc": "2.0",
-                "id": 1,
-                "error": {
-                    "code": -32602,
-                    "message": "too many arguments, want at most 1"
-                }
-            }
-            """
-            if response_json['error']['code'] == -32602:
-                return self.w3.eth.estimateGas({
-                    "from": from_,
-                    "to": to,
-                    "data": data,
-                    "value": Wei(value),
-                })
-            else:
-                raise ValueError(response_json['error'])
-        else:
-            return int(response_json['result'], 16)
+        tx: TxParams = {'to': to}
+        if to:
```

Comment:
Is this one required as you set it in the first step anyways?

Reply:
You are right, thanks